File: packages/whisker-audio/android/src/main/kotlin/rs/whisker/modules/audio/AudioModule.kt
// `whisker-audio` Module (Android). View-less.
//
// Each Rust-side `Player` allocation calls `create(id, source)`, drives
// its entry through `play` / `pause` / `seekTo`, and `release(id)` from
// `PlayerInner::drop` removes it.
//
// Playback state flows back through a single `statusChanged` event whose
// payload carries `playerId`, so the owning Rust runtime can route it
// to the matching handle.

package rs.whisker.modules.audio

import android.os.Handler
import android.os.Looper
import androidx.media3.common.C
import androidx.media3.common.MediaItem
import androidx.media3.common.Player
import androidx.media3.exoplayer.ExoPlayer
import rs.whisker.runtime.RuntimeAttachedListener
import rs.whisker.runtime.Module
import rs.whisker.runtime.WhiskerModule
import rs.whisker.runtime.ModuleDefinition
import rs.whisker.runtime.WhiskerValue


@WhiskerModule
class AudioModule : Module() {

    /**
     * Live players, keyed by the id the Rust side allocates. A plain
     * `HashMap` suffices because every lookup is on the main thread —
     * the Host bridge callback thread is the UI thread on Android.
     */
    private val players: MutableMap<Long, ExoPlayer> = mutableMapOf()
    private val loopFlags: MutableMap<Long, Boolean> = mutableMapOf()

    /**
     * `create` requests that arrived before any `WhiskerView` was attached
     * as a host. Rust's `Player::new` fires from the first render, which
     * runs inside `WhiskerView`'s constructor, so `currentActivity` is
     * still `null` then. Drained by the [RuntimeAttachedListener].
     */
    private val pendingCreates: MutableList<Pair<Long, String>> = mutableListOf()

    /**
     * Installed the first time [pendingCreates] has to queue something,
     * then kept for the process lifetime so the queue drains on every
     * re-attach.
     */
    private var hostListener: RuntimeAttachedListener? = null

    /**
     * Per-player position timer, ticking at ~200 ms while playing so the
     * Rust signal sees smooth progress. Cancelled on pause / stop /
     * release so a paused player doesn't pin a Handler post for nothing.
     */
    private val positionTicker = Handler(Looper.getMainLooper())
    private val tickRunnables: MutableMap<Long, Runnable> = mutableMapOf()

    override fun definition() = ModuleDefinition {
        Name("WhiskerAudio")
        Events("statusChanged")

        Function("create") { args ->
            val id = args.getOrNull(0)?.asInt() ?: return@Function WhiskerValue.Null
            val source = args.getOrNull(1)?.asString() ?: ""
            createPlayer(id, source)
            WhiskerValue.Null
        }
        Function("setSource") { args ->
            val id = args.getOrNull(0)?.asInt() ?: return@Function WhiskerValue.Null
            val source = args.getOrNull(1)?.asString() ?: ""
            players[id]?.let { p ->
                p.setMediaItem(MediaItem.fromUri(source))
                p.prepare()
            }
            WhiskerValue.Null
        }
        Function("play") { args ->
            args.getOrNull(0)?.asInt()?.let { players[it]?.play() }
            WhiskerValue.Null
        }
        Function("pause") { args ->
            args.getOrNull(0)?.asInt()?.let { players[it]?.pause() }
            WhiskerValue.Null
        }
        Function("stop") { args ->
            args.getOrNull(0)?.asInt()?.let { id ->
                players[id]?.let { p ->
                    p.pause()
                    p.seekTo(0)
                }
            }
            WhiskerValue.Null
        }
        Function("seekTo") { args ->
            val id = args.getOrNull(0)?.asInt() ?: return@Function WhiskerValue.Null
            val seconds = args.getOrNull(1)?.asDouble() ?: 0.0
            players[id]?.seekTo((seconds * 1000.0).toLong())
            WhiskerValue.Null
        }
        Function("setVolume") { args ->
            val id = args.getOrNull(0)?.asInt() ?: return@Function WhiskerValue.Null
            val v = (args.getOrNull(1)?.asDouble() ?: 1.0).toFloat().coerceIn(0f, 1f)
            players[id]?.volume = v
            WhiskerValue.Null
        }
        Function("setLoop") { args ->
            val id = args.getOrNull(0)?.asInt() ?: return@Function WhiskerValue.Null
            val looping = args.getOrNull(1)?.asBool() ?: false
            loopFlags[id] = looping
            players[id]?.repeatMode =
                if (looping) Player.REPEAT_MODE_ONE else Player.REPEAT_MODE_OFF
            WhiskerValue.Null
        }
        Function("release") { args ->
            args.getOrNull(0)?.asInt()?.let { id ->
                tickRunnables.remove(id)?.let { positionTicker.removeCallbacks(it) }
                players.remove(id)?.release()
                loopFlags.remove(id)
            }
            WhiskerValue.Null
        }
    }

    /**
     * Create the ExoPlayer for `id` and wire its `Player.Listener` so state
     * changes flow back to Rust. Defers when the host Activity isn't
     * attached yet — the common case during the very first render.
     */
    private fun createPlayer(id: Long, source: String) {
        val ctx = appContext.currentActivity
        if (ctx == null) {
            pendingCreates.add(id to source)
            ensureAttachListener()
            return
        }
        val p = ExoPlayer.Builder(ctx).build()
        if (source.isNotEmpty()) {
            p.setMediaItem(MediaItem.fromUri(source))
            p.prepare()
        }
        p.addListener(object : Player.Listener {
            override fun onIsPlayingChanged(isPlaying: Boolean) {
                dispatchStatus(id)
                if (isPlaying) startTicker(id) else stopTicker(id)
            }
            override fun onPlaybackStateChanged(state: Int) {
                dispatchStatus(id)
            }
            override fun onPlayerError(error: androidx.media3.common.PlaybackException) {
                dispatchStatus(id)
            }
        })
        players[id] = p
        dispatchStatus(id)
    }

    /**
     * Snapshot player `id` and broadcast it as `statusChanged`. No-ops when
     * the player is gone from the map — a listener callback can still land
     * after `release`.
     */
    private fun dispatchStatus(id: Long) {
        val p = players[id] ?: return
        val durationMs = p.duration
        val payload = mapOf(
            "playerId" to WhiskerValue.Int(id),
            "position" to WhiskerValue.Float(p.currentPosition / 1000.0),
            "duration" to WhiskerValue.Float(
                if (durationMs == C.TIME_UNSET) 0.0 else durationMs / 1000.0
            ),
            "isLoaded" to WhiskerValue.Bool(
                p.playbackState == Player.STATE_READY ||
                p.playbackState == Player.STATE_BUFFERING
            ),
            "isPlaying" to WhiskerValue.Bool(p.isPlaying),
        )
        sendEvent("statusChanged", WhiskerValue.Map(payload))
    }

    /**
     * The Handler runs on the main thread, the only one allowed to touch
     * ExoPlayer state.
     */
    private fun startTicker(id: Long) {
        // A rapid pause-play flip would otherwise stack two simultaneous
        // post chains.
        tickRunnables.remove(id)?.let { positionTicker.removeCallbacks(it) }
        val runnable = object : Runnable {
            override fun run() {
                if (players[id] == null) return
                dispatchStatus(id)
                positionTicker.postDelayed(this, 200L)
            }
        }
        tickRunnables[id] = runnable
        positionTicker.post(runnable)
    }

    private fun stopTicker(id: Long) {
        tickRunnables.remove(id)?.let { positionTicker.removeCallbacks(it) }
    }

    /**
     * One-shot install of the listener that drains [pendingCreates].
     * `addOnRuntimeAttachedListener` fires synchronously when a host is
     * already attached, so a late call still lands correctly.
     */
    private fun ensureAttachListener() {
        if (hostListener != null) return
        val listener = RuntimeAttachedListener {
            // Snapshot then clear, so a re-attach mid-drain (rotation)
            // can't see a half-drained queue.
            val pending = pendingCreates.toList()
            pendingCreates.clear()
            for ((id, source) in pending) {
                if (!players.containsKey(id)) {
                    createPlayer(id, source)
                }
            }
        }
        hostListener = listener
        appContext.addOnRuntimeAttachedListener(listener)
    }
}

File: packages/whisker-audio/ios/Sources/WhiskerAudio/AudioModule.swift
// `whisker-audio` Module (iOS). View-less.
//
// Mirrors the Android side: a `[Int64: AVPlayer]` keyed on the
// Rust-allocated id, prop-style functions (`create`, `play`,
// `pause`, `setLoop`, etc.), and a `statusChanged` event whose
// payload carries `{ playerId, position, duration, isLoaded,
// isPlaying }`.
//
// AVPlayer has no built-in loop / status callback API equivalent
// to Media3's `Player.Listener` — looping rides on
// `AVPlayerItemDidPlayToEndTime`, and status updates ride on
// `KVO` (`status`, `rate`) + a periodic time observer at
// ~200 ms cadence.

import AVFoundation
import Foundation
import WhiskerModule
import WhiskerRuntime

public final class AudioModule: Module {

    /// Per-player state. Owned by the module, indexed by the id
    /// the Rust side allocates on `Player::new`.
    private final class PlayerEntry {
        let player: AVPlayer
        var loop: Bool = false
        var endObserver: NSObjectProtocol?
        var statusObservation: NSKeyValueObservation?
        var rateObservation: NSKeyValueObservation?
        /// Foundation Timer driving position dispatch while
        /// playback is active. `addPeriodicTimeObserver` was
        /// unreliable on iOS Simulator (token returned non-nil but
        /// the closure never fired), so we hand-tick instead.
        var progressTimer: Timer?

        init(player: AVPlayer) {
            self.player = player
        }
    }

    private var players: [Int64: PlayerEntry] = [:]

    public override func definition() -> ModuleDefinition {
        ModuleDefinition {
            Name("WhiskerAudio")
            Events("statusChanged")

            Function("create") { (args: [WhiskerValue]) -> WhiskerValue in
                guard let id = args.first?.asInt else { return .null }
                let source = args.count > 1 ? (args[1].asString ?? "") : ""
                self.createPlayer(id: id, source: source)
                return .null
            }
            Function("setSource") { (args: [WhiskerValue]) -> WhiskerValue in
                guard let id = args.first?.asInt,
                      let entry = self.players[id] else { return .null }
                let source = args.count > 1 ? (args[1].asString ?? "") : ""
                self.swapItem(for: entry, id: id, source: source)
                return .null
            }
            Function("play") { (args: [WhiskerValue]) -> WhiskerValue in
                if let id = args.first?.asInt { self.players[id]?.player.play() }
                return .null
            }
            Function("pause") { (args: [WhiskerValue]) -> WhiskerValue in
                if let id = args.first?.asInt { self.players[id]?.player.pause() }
                return .null
            }
            Function("stop") { (args: [WhiskerValue]) -> WhiskerValue in
                if let id = args.first?.asInt, let entry = self.players[id] {
                    entry.player.pause()
                    entry.player.seek(to: .zero)
                }
                return .null
            }
            Function("seekTo") { (args: [WhiskerValue]) -> WhiskerValue in
                guard let id = args.first?.asInt, let entry = self.players[id] else {
                    return .null
                }
                let seconds = args.count > 1 ? (args[1].asDouble ?? 0) : 0
                let time = CMTime(seconds: seconds, preferredTimescale: 600)
                entry.player.seek(to: time)
                return .null
            }
            Function("setVolume") { (args: [WhiskerValue]) -> WhiskerValue in
                guard let id = args.first?.asInt, let entry = self.players[id] else {
                    return .null
                }
                let raw = args.count > 1 ? (args[1].asDouble ?? 1.0) : 1.0
                entry.player.volume = Float(max(0, min(1, raw)))
                return .null
            }
            Function("setLoop") { (args: [WhiskerValue]) -> WhiskerValue in
                guard let id = args.first?.asInt, let entry = self.players[id] else {
                    return .null
                }
                let looping = args.count > 1 ? (args[1].asBool ?? false) : false
                entry.loop = looping
                self.installEndObserver(for: entry, id: id)
                return .null
            }
            Function("release") { (args: [WhiskerValue]) -> WhiskerValue in
                if let id = args.first?.asInt { self.releasePlayer(id: id) }
                return .null
            }
        }
    }

    // MARK: - Player lifecycle

    private func createPlayer(id: Int64, source: String) {
        // iOS audio output is gated on an active AVAudioSession.
        // Without `.playback` + `setActive(true)`, AVPlayer reports
        // rate=1 after `play()` but the audio render loop never
        // pulls samples — currentTime stays at 0 and no audio
        // reaches the speaker. Activating the session is
        // idempotent across players; we re-call on every create
        // so a new player picks up the session even if the host
        // app deactivated it.
        ensureAudioSessionActive()

        let player: AVPlayer
        if !source.isEmpty, let url = URL(string: source) {
            player = AVPlayer(url: url)
        } else {
            player = AVPlayer()
        }
        let entry = PlayerEntry(player: player)
        players[id] = entry
        installKVO(for: entry, id: id)
        installEndObserver(for: entry, id: id)
        dispatchStatus(id: id)
    }

    private func swapItem(for entry: PlayerEntry, id: Int64, source: String) {
        if source.isEmpty {
            entry.player.replaceCurrentItem(with: nil)
        } else if let url = URL(string: source) {
            entry.player.replaceCurrentItem(with: AVPlayerItem(url: url))
        }
        // Loop observer is item-scoped; reinstall on the fresh
        // playerItem so end-of-source seeking still fires when
        // appropriate.
        installEndObserver(for: entry, id: id)
        dispatchStatus(id: id)
    }

    private func releasePlayer(id: Int64) {
        guard let entry = players.removeValue(forKey: id) else { return }
        entry.progressTimer?.invalidate()
        if let token = entry.endObserver {
            NotificationCenter.default.removeObserver(token)
        }
        entry.statusObservation?.invalidate()
        entry.rateObservation?.invalidate()
        entry.player.pause()
        entry.player.replaceCurrentItem(with: nil)
    }

    // MARK: - Observation

    /// Re-broadcast `statusChanged` whenever the AVPlayer's loaded
    /// state or play / pause flag transitions. The rate observer
    /// also starts / stops the position-progress Timer so a
    /// paused player doesn't pin a runloop callback.
    private func installKVO(for entry: PlayerEntry, id: Int64) {
        entry.rateObservation = entry.player.observe(\.rate, options: [.new]) { [weak self] player, _ in
            self?.dispatchStatus(id: id)
            guard let self else { return }
            if player.rate > 0 {
                self.startProgressTimer(for: id)
            } else {
                self.stopProgressTimer(for: id)
            }
        }
        // `status` lives on AVPlayerItem; observe transitions
        // through replaceCurrentItem too — KVO continues across
        // item swaps because we re-look up `.currentItem` inside
        // the closure on every fire.
        entry.statusObservation = entry.player.observe(\.currentItem?.status, options: [.new]) { [weak self] _, _ in
            self?.dispatchStatus(id: id)
        }
    }

    /// Start (or restart) a ~200 ms Timer that re-broadcasts
    /// playback status while the player rate is non-zero.
    ///
    /// We use a hand-rolled Timer instead of
    /// `AVPlayer.addPeriodicTimeObserver` because the latter
    /// silently never fires on iOS Simulator (token comes back
    /// non-nil; the closure is never invoked). Real devices honour
    /// the periodic observer correctly, but the dev loop runs on
    /// the sim so a portable solution wins.
    private func startProgressTimer(for id: Int64) {
        guard let entry = players[id] else { return }
        entry.progressTimer?.invalidate()
        let timer = Timer.scheduledTimer(withTimeInterval: 0.2, repeats: true) { [weak self] _ in
            self?.dispatchStatus(id: id)
        }
        // Also schedule on `.common` modes so a touch-driven runloop
        // mode (e.g. tracking a scroll) doesn't pause the tick.
        RunLoop.main.add(timer, forMode: .common)
        entry.progressTimer = timer
    }

    private func stopProgressTimer(for id: Int64) {
        players[id]?.progressTimer?.invalidate()
        players[id]?.progressTimer = nil
    }

    /// Activate a `.playback` AVAudioSession so AVPlayer's render
    /// loop actually pulls samples. Whisker apps don't currently
    /// take a stance on session category (no recording, mixing,
    /// etc), so `.playback` is the safe default — it solo-routes
    /// audio to the device speaker and survives a screen lock.
    private func ensureAudioSessionActive() {
        let session = AVAudioSession.sharedInstance()
        do {
            try session.setCategory(.playback, mode: .default)
            try session.setActive(true, options: [])
        } catch {
            NSLog("[whisker-audio] AVAudioSession activate failed: \(error)")
        }
    }

    /// (Re-)install the loop-on-end observer. Removes the prior
    /// token before adding a fresh one so a sequence of
    /// `setLoop(true)` / `setLoop(false)` / `setLoop(true)` calls
    /// doesn't stack handlers.
    private func installEndObserver(for entry: PlayerEntry, id: Int64) {
        if let token = entry.endObserver {
            NotificationCenter.default.removeObserver(token)
            entry.endObserver = nil
        }
        guard entry.loop, let item = entry.player.currentItem else { return }
        entry.endObserver = NotificationCenter.default.addObserver(
            forName: .AVPlayerItemDidPlayToEndTime,
            object: item,
            queue: .main
        ) { [weak self] _ in
            self?.players[id]?.player.seek(to: .zero)
            self?.players[id]?.player.play()
        }
    }

    // MARK: - Status payload

    private func dispatchStatus(id: Int64) {
        guard let entry = players[id] else { return }
        let player = entry.player
        let position = CMTimeGetSeconds(player.currentTime())
        let duration = player.currentItem.map { CMTimeGetSeconds($0.duration) } ?? 0
        let safeDuration = duration.isFinite ? duration : 0
        // `currentItem.status == .readyToPlay` is the canonical
        // "loaded" signal, but iOS Simulator hides the
        // transition behind a KVO event that doesn't always
        // fire — derive from "we have a finite duration" instead
        // so the flag flips as soon as the asset metadata
        // resolves, regardless of the status KVO path.
        let isLoaded = safeDuration > 0 || (player.currentItem?.status == .readyToPlay)
        let isPlaying = player.rate > 0 && (player.error == nil)
        let payload: WhiskerValue = .map([
            "playerId": .int(id),
            "position": .float(position.isFinite ? position : 0),
            "duration": .float(safeDuration),
            "isLoaded": .bool(isLoaded),
            "isPlaying": .bool(isPlaying),
        ])
        sendEvent("statusChanged", payload)
    }
}
